src/kernel/log.c

```diff
@@ -0,0 +1,75 @@
+#include <log.h>
+#include <qemu.h>
+#include <framebuffer.h>
+#include <io.h>
+
+char* logLevelStrings[] = {
+    "  [DEBUG] ",
+    "[VERBOSE] ",
+    "   [INFO] ",
+    "  [ERROR] ",
+    "  [FATAL] ",
+};
+const size_t logLevelStrLen = 10; //all the above strings are 10 chars long (excluding null terminator)
+
+size_t logDestBitmap;
+size_t logTrimLevel;
+
+size_t fbCurrentLine;
+size_t fbMaxLine;
+void init_log(size_t defaultOutputs, log_level_t trimBelowLevel){
+    if (defaultOutputs == LOG_OUTPUT_DONT_CARE)
+        defaultOutputs = LOG_OUTPUT_SERIAL; //default to serial
+    
+    logDestBitmap = defaultOutputs;
+    logTrimLevel = trimBelowLevel;
+
+    uint32_t pw, ph, cw, ch;
+    get_framebuffer_mode(&pw, &ph, &cw, &ch);
+    fbMaxLine = ch;
+}
+
+void set_log_trim_level(size_t newTrim){
+    logTrimLevel = newTrim;
+}
+
+void logln(log_level_t level, char* msg){
+    if (level < logTrimLevel)
+        return; //dont log things that we dont want to see for now. (would be nice to store these somewhere in the future perhaps, just not display them?)
+    
+    for (size_t i = 0; i < LOG_OUTPUT_COUNT; i++){
+        if ((logDestBitmap & (1 << i)) == 0)
+            continue; //bit is cleared, we should not log there
+
+        switch (1 << i){
```

Comment:
do we want to support the legacy vga case too?

And in case do we want to make a new case, or support it inside the framebuffer one?

Reply:
Sure I can add it :) since the system likely won't support running both the mmio framebuffer and vga at the same time, I'd say we make them the same case. That way you just select if you want it to output to the screen or not,rather than thinking about how it gets there.